File: CodingSamples/Fundamentals/Generic/TypeParamTest/DemoApp2/Common/SimpleStack.cs
using System.Linq.Expressions;

namespace Common;

//a generic class with type parameter E
public class SimpleStack<E> : IStackReader<E>, IStackWriter<E>
{
    //nested class
    class Node
    {
        internal E element;
        internal Node below;
    }

    private Node top;

    public int Length { get; private set; }

    public void Push(E item)
    {
        top = new Node { element = item, below = top }; //instance initializer
        Length += 1;
    }

    public E Pop()
    {
        E item = top.element;
        top = top.below;
        Length -= 1;
        return item;
    }
}

File: CodingSamples/Java/Fundamentals/Generic/IterableTest/demo1/common/SimpleStack.java
package common;

import java.util.Iterator;

public class SimpleStack<V>  implements Iterable<V> {

    //inner member class
    class Node {

        V element;
        Node below;

        Node(V item){
            element = item;
            below = top;
        }
    }

    private Node top;

    public void push(V item) {
        top = new Node(item);
    }

    public V pop() {
        V item = top.element;
        top = top.below;
        return item;
    }

    public boolean empty() {
        return top == null;
    }

    public Iterator<V> iterator() {
        //instantiating an inner local anonymous class 
        //that implements Iterator<V>
        return new Iterator<V>(){

            private Node current = top;

            public boolean hasNext() {
                return current != null;
            }

            public V next() {
                V result = current.element;
                current = current.below;
                return result;
            }
        };
    }

}

